src/main/java/com/bpodgursky/jbool_expressions/Expression.java

```diff
@@ -6,21 +6,14 @@
 
 public abstract class Expression<K> implements Comparable<Expression> {
 
-  public long hash;
-
   public int compareTo(Expression o) {
-    return Long.compare(this.hash, o.hash);
-  }
-
-  @Override
-  public boolean equals(Object o){
-    return o instanceof Expression &&
-           this.hash == ((Expression) o).hash;
-  }
-
-  @Override
-  public int hashCode() {
-    return (int) this.hash;
+    int compare = Integer.compare(hashCode(), o.hashCode());
+    if(compare == 0 && !equals(o)) {
+      // If hashcode matches and expressions are not equal then we may have a hash collision.
+      // This is very unlikely to happen but if it does then go for string comparison (slow).
+      return toString().compareTo(o.toString());
```

Comment:
this creates an odd ordering (first by hash, then by string order) but I don't really see anything wrong with it.  With 32 bit hashes we definitely have to handle collisions though.